liquid-staking/src/delegation.rs

```diff
@@ -0,0 +1,163 @@
+                current_epoch,
+            ))
+            .async_call_and_exit();
+    }
+
+    #[callback]
+    fn unstake_callback(
+        &self,
+        caller: &ManagedAddress,
+        egld_amount_to_unstake: &BigUint,
+        current_epoch: Epoch,
+        #[call_result] result: ManagedAsyncCallResult<()>,
+    ) {
+        match result {
+            ManagedAsyncCallResult::Ok(()) => {
+                self.delegated_value(caller)
+                    .update(|value| *value -= egld_amount_to_unstake);
+                self.undelegate_epoch(caller)
+                    .set(current_epoch + UNBOND_PERIOD);
+            }
+            _ => sc_panic!("There was an error at delegating"),
+        }
+    }
+
+    #[endpoint(claimRewardsFromDelegation)]
+    fn claim_rewards_from_delegation(&self, contracts: MultiValueEncoded<ManagedBuffer>) {
+        let caller = self.blockchain().get_caller();
+        self.require_caller_has_stake(&caller);
+
+        for delegation_contract in contracts {
+            let delegation_mapper = self.delegation_addresses(&delegation_contract);
+            if !delegation_mapper.is_empty() {
+                let delegation_address = delegation_mapper.get();
+                self.tx()
+                    .to(delegation_address)
+                    .raw_call(CLAIM_REWARDS_ENDPOINT)
+                    .callback(DelegationModule::callbacks(self).claim_rewards_from_delegation_cb())
+                    .async_call_and_exit();
+            }
+        }
+    }
+
+    #[callback]
+    fn claim_rewards_from_delegation_cb(
+        &self,
+        // egld_amount: &BigUint,
+        #[call_result] result: ManagedAsyncCallResult<()>,
+    ) {
+        match result {
+            ManagedAsyncCallResult::Ok(()) => {}
+            _ => sc_panic!("There was an error at claiming rewards from the delegation contract"),
+        }
+    }
+
+    // NOTE: Should this also add to the map ?
+    #[endpoint(slashValidator)]
+    fn slash_validator(
+        &self,
+        validator_address: ManagedAddress,
+        bls_key: ManagedBuffer,
+        value_to_slash: BigUint,
+    ) {
+        let caller = self.blockchain().get_caller();
+        self.require_caller_to_be_header_verifier(&caller);
+        self.require_bls_key_to_be_registered(&bls_key);
+
+        require!(
+            !self.validator_bls_key_address_map(&bls_key).is_empty(),
+            "There is no associated address to the given BLS key"
+        );
+
+        self.require_caller_has_stake(&validator_address);
+
+        require!(value_to_slash > 0, "You can't slash a value of 0 eGLD");
+
+        let delegated_value = self.delegated_value(&validator_address).get();
+        require!(
+            delegated_value >= value_to_slash,
+            "The slash value can't be greater than the total delegated amount"
+        );
+
+        self.delegated_value(&validator_address)
+            .update(|value| *value -= &value_to_slash);
```

Comment:
Slight optimization to only build the key once:

```suggestion
        let mapper = self.delegated_value(&validator_address).get();
        let delegated_value = mapper.get();
        require!(
            delegated_value >= value_to_slash,
            "The slash value can't be greater than the total delegated amount"
        );

        
           mapper.update(|value| *value -= &value_to_slash);
```

Reply:
Also, consider using `AddressToIdMapper` where you have address as part of the key. Any storage key over 32 bytes costs extra gas to use.

Reply:
Example: https://github.com/multiversx/sc-guilds-rs/blob/5b4f8eba503a173d2fbb6f110d33c79d4069e807/guild-factory/src/factory.rs#L191

And then, you simply have `fn delegated_value(&self, address_id: AddressId) -> SingleValueMapper<BigUint>;`

Reply:
Noted as TODO